src/java.base/share/classes/java/lang/Runtime.java

```diff
@@ -83,10 +83,10 @@ private Runtime() {}
      * until they finish.  Once this is done the virtual machine
      * {@linkplain #halt halts}.
      *
-     * <p> If this method is invoked after all shutdown hooks have already
-     * been run and the status is nonzero then this method halts the
-     * virtual machine with the given status code. Otherwise, this method
-     * blocks indefinitely.
+     * <p> If this method is invoked from more than one thread the status code
+     * of the first invocation will be used; the status codes from other invocations
+     * will be ignored. If this method is invoked from a shutdown hook the system
+     * will deadlock.
```

Comment:
Expressing this accurately is tricky - what is "first" here? I suggest the following:

> Invocations of this method are serialized such that only one invocation will actually proceed with the shutdown sequence and terminate the VM with the given status code. All other invocations will block indefinitely. If this method is invoked from a shutdown hook the system will deadlock.

Reply:
I reworded with the suggested text in mind. I also added another sentence referencing native signal handlers, which may also initiate shutdown. I think this clarification is important so that it is clear the status code of all invocations from Java may still be ignored. See [2253259](https://github.com/openjdk/jdk/pull/9351/commits/2253259c82b13e7b2186429a80cc42497235b035).

Reply:
I think the wording in the latest commit (9d972b) is problematic. One reason is that you've changed it to "will run shutdown hooks" but it doesn't run the hooks, it starts them, and so conflicts with the previous paragraph. I also think "That invocation may be initiated via platform specific signal handlers" is confusing here as there is no notion of "signal handler" to reference. There may be scope for text elsewhere, maybe with an implNote for signals such as HUP, but I don't think this is the PR for this. So I think it better to try the wording that David suggested and see if it needs any improvement.

Reply:
YAO (Yet Another Opinion)!  But I think we're actually converging.

David's wording:
>Invocations of this method are serialized such that only one invocation will actually proceed with the shutdown sequence and terminate the VM with the given status code.

... gives the impression that an invocation of this method WILL terminate the VM with the given status code - which is not actually true, given the potential for signals. This is already alluded to in `Runtime::addShutdownHook`. I agree that this could be resolved separately, but we should _allow_ for it.
 
The original wording:
> If this method is invoked after all shutdown hooks have already been run ... 

... seems quite clever (and allows for signals). Maybe we could:
> If this method is invoked after shutdown hooks have already been started it will block indefinitely. If this method is invoked from a shutdown hook the system will deadlock.

Reply:
I like the new suggested wording, but I would slightly tweak it. As Alan mentioned in another comment the first paragraph makes clear the method never returns normally. How does the following sound?

> If this method is invoked after shutdown hooks have already been started, the supplied status code will be ignored. If this method is invoked from a shutdown hook the system will deadlock.

Reply:
> gives the impression that an invocation of this method WILL terminate the VM with the given status code - which is not actually true, given the potential for signals.

The subtle distinction being calling Runtime.exit versus Shutdown.exit. The actual serialization takes places in Shutdown.exit and serializes with signals too. But discussion of signals has no place here - I'm not even sure we document those ways of terminating the VM anywhere?

Reply:
@dholmes-ora Agreed - this paragraph should not explicitly mention signals or any _other_ means of termination.  

There is some shutdown hook specific verbiage relating to termination in `Runtime::addShutdownHook` :

```
In rare circumstances the virtual machine may abort, that is, stop running without shutting
down cleanly. This occurs when the virtual machine is terminated externally, for example
with the SIGKILL signal on Unix or the TerminateProcess call on Microsoft Windows. The
virtual machine may also abort if a native method goes awry by, for example, corrupting
internal data structures or attempting to access nonexistent memory. If the virtual machine
aborts then no guarantee can be made about whether or not any shutdown hooks will be run.
```

... but we do not need to go there, and it's not really enough to leverage anyway. 

>The subtle distinction being calling Runtime.exit versus Shutdown.exit. The actual serialization takes places in Shutdown.exit and serializes with signals too.

Exactly. On further reading of the proposed "Invocations of this method are serialized such that only one invocation will actually proceed with the shutdown sequence and terminate the VM with the given status code.", I think that it is fine (given its position and context in Runtime::exit). My reason for suggesting an alternative ("If this method is invoked after shutdown hooks have already been started, it will block indefinitely") was mainly to try to find something a little simpler while retaining correctness.  Given the discussion so far, either is fine with me.